File: js/src/chunker/chunker.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import { chunkText } from "./chunker.js";

// Proven against the shared fixture — every case must match the Python reference
// citenexus.evidence.chunker.chunk_text exactly. Follows the §4 exemplar: load
// conformance/cases/chunker.json, assert deep equality over ALL cases, no leniency.
interface ChunkerCase {
  text: string;
  max_tokens: number;
  overlap: number;
  chunks: string[];
}

/** Vector counts, pinned. A vector silently dropped from the fixture is a
 *  weakened contract that no per-case assertion can see. */
const EXPECTED_COUNTS: Record<string, number> = { cases: 7 };

describe("chunker conformance", () => {
  const cases = loadCase<ChunkerCase[]>("chunker.json");

  it("vector counts are pinned", () => {
    expect({ cases: cases.length }).toEqual(EXPECTED_COUNTS);
  });

  for (const [i, c] of cases.entries()) {
    it(`chunkText case ${i} (max=${c.max_tokens}, overlap=${c.overlap})`, () => {
      expect(chunkText(c.text, c.max_tokens, c.overlap)).toEqual(c.chunks);
    });
  }
});

File: golang/chunker/chunker_test.go
package chunker

import (
	"reflect"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// expectedChunkerCases pins the size of conformance/cases/chunker.json.
const expectedChunkerCases = 7

func TestChunkerVectorCount(t *testing.T) {
	var cases []struct {
		MaxTokens int `json:"max_tokens"`
	}
	conform.Case(t, "chunker.json", &cases)
	if len(cases) != expectedChunkerCases {
		t.Fatalf("chunker.json: got %d cases, want %d", len(cases), expectedChunkerCases)
	}
}

// The chunker is proven against the shared fixture — every case must match the
// Python reference exactly (SPEC-PORTS-v1 §4/§10). Byte-identical output over ALL
// cases, no leniency, no skips.
func TestChunkerConformance(t *testing.T) {
	var cases []struct {
		Text      string   `json:"text"`
		MaxTokens int      `json:"max_tokens"`
		Overlap   int      `json:"overlap"`
		Chunks    []string `json:"chunks"`
	}
	conform.Case(t, "chunker.json", &cases)

	if len(cases) != expectedChunkerCases {
		t.Fatalf("chunker.json: got %d cases, want %d", len(cases), expectedChunkerCases)
	}
	for i, c := range cases {
		got := ChunkText(c.Text, c.MaxTokens, c.Overlap)
		want := c.Chunks
		if want == nil {
			want = []string{}
		}
		if !reflect.DeepEqual(got, want) {
			t.Fatalf("case %d: ChunkText(%q, max=%d, overlap=%d) =\n  %#v\nwant\n  %#v",
				i, c.Text, c.MaxTokens, c.Overlap, got, want)
		}
	}
}
